enhancements/authentication/direct-oidc-configuration.md

```diff
@@ -0,0 +1,457 @@
+invalid.
+
+Validation rules for the API:
+1. The `OIDCProviders` field can only be set iff the `Type` field is set to the `OIDC`
+   value
+2. Only a single `OIDCProvider` can ever be set. This requirement can be lifted
+   once upstream allows setting multiple OIDC IdPs in its structured authentication
+   configuration.
+3. The `OIDCProvider.Issuer.URL` MUST always use the "https://" scheme
+4. Only a single audience can ever be set in `OIDCProvider.Issuer.Audiences`. This
+   requirement can be lifted once upstream allows setting multiple audiences
+   in its structured authentication configuration.
+5. If `OIDCProvider.Issuer.CertificateAuthority` is set, it MUST refer an existing
+   configMap that contains the `ca.crt` key in its `.data` field.
+
+### Implementation Details
+
+The following implementation describes how to implement the feature in a standalone
+OCP. The configuration API might be different in other derived platforms, but aside
+from that all the other implementation details are binding for all such platforms.
+
+This feature gets configured by setting `Type: OIDC` in the `authentications.config.openshift.io/cluster`
+resource. This is a sign for the operator handling the cluster authentication stack
+to remove the OpenShift integrated authentication by removing the integrated
+oauth-server and the oauth-apiserver.
+
+The removal of the oauth-apiserver means that the `user.openshift.io` and
+`oauth.openshift.io` APIs will be unavailable. As this would leave dangling objects
+in etcd, all objects of these APIs should be removed prior to the API removal.
+
+In an initial implementation, this removal shall be manual and performed by the user.
+For GA, the operator handling the oauth-apiserver deployment should turn the API
+server into a mode that refuses to accept any new CREATE requests on these APIs,
+after which the objects get removed by the operator maintaining the oauth-apiserver.
+
+> [!NOTE]
+> TODO: will this break the Projects API?
+
+The platform MUST provide means of administrative user authentication even without
+the oauth-server. In case of standalone OpenShift, this is provided by the
+`system:admin` certificate/key pair present in the installation kubeconfig.
+
+Once `Type: OIDC` is set and the configuration passes validation, the operator
+configuring the kube-apiserver converts the API values into appropriate flags.
+In later versions, when the underlying kube-apiserver implements [a stable structured
+authentication configuration](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-authentication-configuration),
+the values from the API get transformed into this new structure.
+
+#### Hypershift
+
+Hypershift is not prepared today for Tech Preview features and does not appear
+to have a way to mark cluster support to be limited as such. However, this feature
+in its first iteration will not be suitable for production.
```

Comment:
We need to clarify expectations for the feature stability. We have the means to hide features and ship only in targeted controlled environments. But I assume those environments would expect the feature commitment to be prod ready, not tech preview.
Also If we change our API with something subject to have backward incompatible changes  this might impact our ability to release in self hosted hcp. Unless you are saying behaviour might change but API must not.

Reply:
The only feature stability expectation that I can provide is that the core workloads should work. I assume 3rd party applications that provide API with OpenShift-based authentication will likely be broken. Hence Tech Preview to identify those.

The API proposed in this API should become stable once implemented. The behavior will likely, too. Third party integrations will break because we cannot influence them.